### PR TITLE
web: search: major code excerpt refactor (better testing, code structure, etc.)

### PR DESCRIPTION
**Pull request by slimsag**
## Overview

This is a prerequisite requirement for implementing optimizations like #6992 and #16467 in a clean fashion.

It refactors how `CodeExcerpt`, `FileMatchChildren`, and `FileMatch` work together, giving us:

- Better, more isolated unit-testing of the functionality which breaks line matches up into groups.
- More tests (added in this PR.)
- A more clear separation of concerns.

## The problem

The code structure behind `CodeExcerpt` is a bit spaghetti today.

1. (ok) `FileMatch` takes the actual individual-line matches returned from the search GraphQL API and passes those to `FileMatchChildren`.
2. (ok) `FileMatch` tells `FileMatchChildren` exactly how many matches (`FileMatchProps.subsetMatches`) to show when the file is collapsed (user has not clicked "Show N more matches").
3. (ok) `FileMatchChildren` groups the individual line matches into groups based on:
    - The number of desired matches to show (`FileMatchProps.subsetMatches`)
    - Whether or not the matches are contiguous regions or not.
4. (ok) `FileMatchChildren` renders a `CodeExcerpt` for each group of matches it wants rendered.
5. (?) `FileMatchChildren` gives the `CodeExcerpt` a function to fetch the *entire* highlighted file instead of giving it the subset of matches and syntax highlighted lines it is responsible for.
6. (??) `CodeExcerpt` must calculate how many lines it should display which [is very complex](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@55d41b342736b318b76cb5ffe54db14ed5558ee6/-/blob/client/shared/src/components/CodeExcerpt.tsx#L120-156).
7. (???) `CodeExcerpt` must account for the user-configurable number of context lines, instead of rendering the section of code the parent told it.
8  (???!) `CodeExcerpt` must account for adjacent `CodeExcerpt`s

In short, the actual logic for how we go from "individual line matches -> groups of matches" is today split across multiple locations:

- `FileMatchChildren`
- `CodeExcerpt`
- `FileMatchContext`

A change in any of the three can break the behavior in totality, and it is quite brittle / hard to test.

## Solution

This change lifts the logic for "individual line matches -> groups of matches" out of the three components and into a single function:

```ts
/**
 * Calculates how to group together matches for display. Takes into account:
 *
 * - Whether or not displaying a subset or all matches is desired
 * - A surrounding number of context lines to display
 * - Whether or not context lines have (or do not have) matches within them
 * - Grouping based on whether or not there is overlapping or adjacent context.
 *
 * @param matches The matches to split into groups.
 * @param maxMatches The maximum number of matches to show, or 0 for all.
 * @param context The number of surrounding context lines to show for each match.
 */
export const calculateMatchGroups = (matches: IMatchItem[], maxMatches: number, context: number): IGroupedMatches
```

Thus making the behavior of the components here very straightforward:

- `FileMatch` renders a file match by creating `FileMatchChildren`.
- `FileMatchChildren` takes the search results and sends them to `calculateMatchGroups`, rendering a `CodeExcerpt` for each group.
- Each `CodeExcerpt` is now a simple, straightforward component only responsible for the excerpt of code _it is rendering_.
    - No longer responsible for splitting the excerpt of code it is rendering out of the entire file.
    - No longer responsible for caring about adjacent excerpts of code and performing complex end-of-excerpt calculations.
    - No longer is the parent partially responsible for splitting, and passing "fun" parameters like this to the `CodeExcerpt`:
        ```
            /** The highest line number among the subset matches. */
            lastSubsetMatchLineNumber: number
        ```

## Future improvements & confidence in this change

I think `calculateMatchGroups` can be made a lot more straightforward in how it operates. The logic is sound, but the way it is written is convoluted. I initially tried to tackle this, but due to the lack of tests and isolation I ended up wasting ~7d here attempting to rewrite this logic and discovering numerous edge cases that needed to be handled when doing side-by-side comparisons.

This change is mostly just moving the logic that previously existed into another location, so there are zero behavior changes here and I have done extensive automated, and manual, testing to ensure that is the case.

Now that this code is isolated and heavily tested, we should be able to do some more fearless refactoring of it to make it more legible.

Helps #6992 and #16467

Signed-off-by: Stephen Gutekanst <stephen.gutekanst@gmail.com>

### CODE REVIEW

**Review by limitedmage — APPROVED** [review]

File: client/shared/src/components/FileMatchContext.ts [hunk 1]
```diff
@@ -22,3 +26,159 @@ export const mergeContext = <T extends { line: number }>(context: number, highli
...
+ * Describes matches that have been grouped together.
+ */
+interface IGroupedMatches {
+    // The grouped matches.
+    grouped: IGroupedMatch[]
+
+    // The subset of sorted matches that should be displayed. Useful if e.g. rendering matches as
+    // plaintext.
+    matches: IMatchItem[]
+}
```
Comment (rvantonder) [inline client/shared/src/components/FileMatchContext.ts:120]: Nit: the type containing `matches: IMatchItem[]` doesn't really adhere to 

> Describes matches that have been grouped together.

It's more like a convenient format of flat, sorted matches. I would probably avoid this bundling and just return a "tuple" from the function directly like `: [IMatchItem[], IMatchGroups[]]`. But even better it also looks like `calculateMatchGroups` can be decoupled to a function that returns `MatchItem[]` (just the sort operation) and then a second function that calculates groups from this input. I don't see an obvious dependency that force these two operations to happen in the same function, except that they share some params (`maxMatches` and `context`). So I would share the param signature across the two functions instead (you could bundle them in a type for match options). The caller will have to be updated to accordingly (call the second function on the result of the first function for groups where needed, or you can create a convenience function that returns both if that's really needed in every caller, but it looks like there's only one).

**Review by rvantonder — APPROVED** [review]
It's the refactor tractor

File: client/shared/src/components/FileMatchContext.ts [hunk 2]
```diff
@@ -22,3 +26,159 @@ export const mergeContext = <T extends { line: number }>(context: number, highli
...
+            startLine,
+            endLine: endLine + 1,
+        }
+    }
+}
+
+/**
+ * Describes a single group of matches.
+ */
+interface IGroupedMatch {
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply client/shared/src/components/FileMatchContext.ts:87]: Also the thing @limitedmage said about Hungarian notation

File: client/shared/src/components/CodeExcerpt.test.tsx [hunk 3]
```diff
@@ -43,27 +44,17 @@ describe('CodeExcerpt', () => {
...
-        const { container } = render(
-            <CodeExcerpt
-                {...defaultProps}
-                context={5}
-                highlightRanges={[{ line: 4, character: 1, highlightLength: 2 }]}
-                lastSubsetMatchLineNumber={4}
-            />
-        )
-        expect(container.querySelectorAll('.code-excerpt tr').length).toBe(10)
+        expect(container.querySelectorAll('.code-excerpt tr').length).toMatchInlineSnapshot('3')
```
[the remark this answers is not in the document]
Comment (slimsag) [reply client/shared/src/components/CodeExcerpt.test.tsx:57]: Ah, I just found this easier because yarn can auto-update snapshots - but happy to change this.

```suggestion
        expect(container.querySelectorAll('.code-excerpt tr').length).toBe(3)
```

[the remark this answers is not in the document]
Comment (limitedmage) [reply client/shared/src/components/CodeExcerpt.test.tsx]: I find snapshots useful for big things like Enzyme-mounted React components, but small things like this I think it's better to force you to manually update it.

File: client/shared/src/components/FileMatchChildren.tsx [hunk 4]
```diff
@@ -124,47 +89,60 @@ export const FileMatchChildren: React.FunctionComponent<FileMatchProps> = props
...
+                            repoName={props.result.repository.name}
+                            commitID={props.result.file.commit.oid}
+                            filePath={props.result.file.path}
+                            startLine={group.startLine}
+                            endLine={group.endLine}
+                            highlightRanges={group.matches}
+                            className="file-match-children__item-code-excerpt"
+                            isLightTheme={props.isLightTheme}
+                            /* eslint-disable react/jsx-no-bind */
+                            fetchHighlightedFileRangeLines={() =>
```
[the remark this answers is not in the document]
Comment (slimsag) [reply client/shared/src/components/FileMatchChildren.tsx:114]: Got it, I figured this was the only way to do this because I need to access the contextual information `group` for the code excerpt being rendered and saw this pattern used in a few other places - I am not familiar with `useCallback` - could you give me some more pointers/tips on how I might refactor this? (happy to give it a shot either way)

File: client/shared/src/components/FileMatchContext.ts [hunk 5]
```diff
@@ -22,3 +26,159 @@ export const mergeContext = <T extends { line: number }>(context: number, highli
...
+        character: number
+        highlightLength: number
+        badge: BadgeAttachmentRenderOptions | undefined
+        IsInContext: boolean
+    }[],
+    context: number,
+    highestLineNumberWithinSubsetMatches: number
+): IGroupedMatch => {
+    {
+        const contextLines = context || context === 0 ? context : 1
```
[the remark this answers is not in the document]
Comment (slimsag) [reply client/shared/src/components/FileMatchContext.ts:42]: lol, I copied this 1:1 over to here / didn't write it myself. I agree it is bonkers and cannot evaluate to `1`. Will fix.

File: client/shared/src/components/FileMatchChildren.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (limitedmage) [reply client/shared/src/components/FileMatchChildren.tsx:114]: You'd have something like this (eyeballing it here, linter will help) near the top of the function:
```ts
const fetchHighlightedFileRangeLines = useCallback((startLIne, endLIne) => props
                                    .fetchHighlightedFileLines(
                                        {
                                            repoName: props.result.repository.name,
                                            commitID: props.result.file.commit.oid,
                                            filePath: props.result.file.path,
                                            disableTimeout: false,
                                            isLightTheme: props.isLightTheme,
                                        },
                                        false
                                    )
                                    .pipe(map(lines => lines.slice(startLine, endLine))),
[props.result, props.isLightTheme, props.fetchHighlightedFileLines])
```

and you pass it to the `CodeExcerpt` here. You'd need to modify `CodeExcerpt` to pass the `startLIne` and `endLIne` to the function it gets as well.

`useCallback` will cache the function so it's not recreated every time the element renders, only if there's a change based on the dependency list.

Let me know if you need more help.

File: client/shared/src/components/FileMatchChildren.tsx [hunk 6]
```diff
@@ -54,60 +60,19 @@ export const FileMatchChildren: React.FunctionComponent<FileMatchProps> = props
...
-            : 0
-
-    const showItems = sortedItems.filter(
-        (item, index) =>
-            props.allMatches ||
-            index < props.subsetMatches ||
-            item.line <= highestLineNumberWithinSubsetMatches + context
-    )
+    const maxMatches = props.allMatches ? 0 : props.subsetMatches
+    const { matches, grouped } = calculateMatchGroups(props.items, maxMatches, context)
```
Comment (limitedmage) [inline client/shared/src/components/FileMatchChildren.tsx:64]: Thinking maybe this should be wrapped in a `useMemo`, how expensive is this?

File: client/shared/src/components/FileMatchChildren.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (limitedmage) [reply client/shared/src/components/FileMatchChildren.tsx:114]: Here's an example of `useCalllback` in our code: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@jp/streaminsearchqa2/-/blob/client/web/src/search/results/streaming/StreamingSearchResultsFilterBars.tsx#L42:44

File: client/shared/src/components/FileMatchChildren.tsx (hunk 6 above)
Comment (slimsag) [reply client/shared/src/components/FileMatchChildren.tsx:64]: Good call, I don't think it is super expensive but probably worth memoizing

File: client/shared/src/components/FileMatchContext.ts (hunk 1 above)
Comment (slimsag) [reply client/shared/src/components/FileMatchContext.ts]: I tried to do this just now but it's a bit more complex than expected: `matches: IMatchItem[]` is not just the sorted list but also just the _shown_ matches (determined based on context and other input parameters). If I split this out into another function, then the 2nd function needs as inputs: those `matches` and `highestLineNumberWithinSubsetMatches` or the original `sortedMatches`.

I think we can do _something_ like this, but it feels a bit heavy to tack onto this PR right now so I will go the "tuple" approach you mentioned instead

File: client/shared/src/components/FileMatchChildren.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (slimsag) [reply client/shared/src/components/FileMatchChildren.tsx]: This was super helpful, thank you a ton for the guidance here!

**Comment by slimsag** [comment]
@limitedmage could you take a quick look at my usage of `useMemo` and `useCallback` here? everything looks good and passes tests but want to make sure I didn't miss anything :)

**Review by limitedmage — APPROVED** [review]
New changes look good!

File: client/shared/src/components/FileMatch.tsx [hunk 7]
```diff
@@ -12,22 +12,25 @@ import { BadgeAttachmentRenderOptions } from 'sourcegraph'
 
 const SUBSET_COUNT_KEY = 'fileMatchSubsetCount'
 
-export type IFileMatch = Partial<Pick<GQL.IFileMatch, 'revSpec' | 'symbols' | 'limitHit'>> & {
+export type FileLineMatch = Partial<Pick<GQL.IFileMatch, 'revSpec' | 'symbols' | 'limitHit'>> & {
```
Comment (felixfbecker) [inline client/shared/src/components/FileMatch.tsx:15]: This seems to have been an old workaround to deep-pick a subset of a GraphQL type. Since we now have auto-generated types, could we use that?

Comment (slimsag) [reply client/shared/src/components/FileMatch.tsx:15]: @felixfbecker to be honest I am not sure what this means:

> Since we now have auto-generated types, could we use that?

This PR has already become quite large and the change here was just to ensure we don't use Hungarian notation as was suggested to me here https://github.com/sourcegraph/sourcegraph/pull/16466#discussion_r537767126

Comment (slimsag) [reply client/shared/src/components/FileMatch.tsx:15]: Gonna merge this as-is now to unblock my other work which depends on this, but if you give me some more hints about what the change would be here I'm happy to follow up with another PR for this later

Comment (felixfbecker) [reply client/shared/src/components/FileMatch.tsx:15]: Sorry, I thought I had linked to the docs! https://docs.sourcegraph.com/dev/background-information/web/graphql

This is not a blocking issue, just a suggestion as I saw you were doing code cleanup in this PR, this would be another major cleanup (and increased type safety, maintainability etc). Basically you don't need to define these GraphQL type subsets manually anymore, you can use a type from a GraphQL query or a GraphQL fragment that is used to fetch the data. You could define the GraphQL fragment here for example and export it and use it in the GraphQL query, while also using the auto-generated TypeScript types for that fragment in the rest of the file.
Again - no blocker.

Comment (slimsag) [reply client/shared/src/components/FileMatch.tsx:15]: Ahhh, I see, very interesting, thank you for the additional info! I will see about making this change later this week - sounds promising
